### PR TITLE
network: fixing a bug in detecting closed connections

### LINKED ISSUE #3785: flake in RouterDownstreamDisconnectBeforeRequestComplete
I've now seen this in uds_integration_test on asan, and http2_integration_test.

IpVersions/Http2IntegrationTest.RouterDownstreamDisconnectBeforeRequestComplete/IPv6
IpVersions/IntegrationTest.RouterDownstreamDisconnectBeforeRequestComplete/IPv4
TestParameters/UdsUpstreamIntegrationTest.RouterDownstreamDisconnectBeforeRequestComplete/0

I guess that rules out a uds bug.

### PR DESCRIPTION
**Pull request by alyssawilk**
As documented in #3785, we had a disconnect test flaking out because the ECONNRESET was being overwritten by EAGAIN in the logging code, so the ECONNRESET was not properly handled and the test hung.  This represents an actual Envoy bug where an inopportune log flush can mask a connection close.    Presumably the same issue exists on the write path.

*Risk Level*: Low
*Testing*: IpVersions/Http2IntegrationTest.RouterDownstreamDisconnectBeforeRequestComplete/IPv6 less flaky
*Docs Changes*: n/a
*Release Notes*: n/a
Fixes #3785

### CODE REVIEW

File: source/common/network/raw_buffer_socket.cc [hunk 1]
```diff
@@ -18,6 +18,7 @@ IoResult RawBufferSocket::doRead(Buffer::Instance& buffer) {
   do {
     // 16K read is arbitrary. TODO(mattklein123) PERF: Tune the read size.
     int rc = buffer.read(callbacks_->fd(), 16384);
+    int error = errno; // Latch errno before any logging calls can overwrite it.
```
Comment (htuch) [inline source/common/network/raw_buffer_socket.cc]: Nit: make these all `const`, less likely that we will forget it after latching ;)

**Comment by htuch** [comment]
Can we fix this structurally? E.g. in our formatter, forbid the use of `errnor` except in assignment?

**Comment by alyssawilk** [comment]
Yeah, our test only caught the bug in doRead - I found the others via visual inspection grepping about for errno.   ClientConnectionImpl::connect looks sketchy but it is legit insofar as there weren't any other syscalls (even via sneaky log macros) and the rest looked fine to me.

Unfortunately, grepping isn't terribly futureproof - if you'd prefer I could blanket latch errno to error every time we reference it even in currently benign cases.    internally we go as far as plumb an errno int* through to our syscall wrappers to avoid issues if folks add debug logging in the wrappers.

File: source/common/network/raw_buffer_socket.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/network/raw_buffer_socket.cc]: Darn it!  
Days since alyssar@ forgot const checks -> 0 :-P

**Comment by alyssawilk** [comment]
At compile time?  I have no clue how to fix it.  We could do best-effort fix_format checks that you can't have LOG in the same line as errno but that's pretty weak.  You could have an ENVOY_LOG("syscall failed") between syscall and errno checking and it'd pass both those checks so neither is really sufficient.
Offhand we could do a horrible hack where in logging in debug binaries we explicitly clear errno in debug log macros so anyone testing for error handling would catch their screw ups fast.    I think that would have caught two and maybe all three of these.

**Comment by mattklein123** [comment]
The issue with any hacks is that it doesn't prevent someone from doing some other non-logging system call that resets errno. IMO this is just errno terribleness. As long as you search for errno and this is what you found this seems good to me. Scary stuff though. Up to you if you want to do something more.

**Comment by htuch** [comment]
I mean, wouldn't it be enough to do something like `ag errno | ag -v "err = errno" | ag -v "errno = E" | ag -v "errno:" | ag -v <some exclusion for comments>` in `check_format`?

**Comment by alyssawilk** [comment]
htuch@ assuming ag is some grep alias, you're suggesting we force-require latching errno to error everywhere?  That probably gets us halfway there - I still had to do code inspection for the readv syscalls which were buried 3 deep in function stacks to make sure there were no routes from syscall to latching that were dangerous and that work isn't futureproof.    I don't think the fix format scripts are worth it unless we also do the API cleanup to latch errno directly next to the system call and changing your theoretical required latching to something like
int latching_errno_right_after_syscall = errno; 
Might be a nice standalone for your upcoming noogler?

**Pull request merged**